DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/README.md

```diff
@@ -0,0 +1,206 @@
+
+| File                           | Description 
+|:---                            |:---
+|`merge_sort.cpp`                | Contains the `main()` function and the top-level interfaces.
+|`MergeSort.hpp`                 | The function to submit all of the merge sort kernels (`Partition`, `Produce`, `Merge`, and `Consume`).
+|`Consume.hpp`                   | The `Consume` kernel for the merge unit. This kernel reads from an input pipe and writes out to either a different output pipe, or to device memory.
+|`Merge.hpp`                     | The `Merge` kernel for the merge unit and the merge tree. This kernel streams in two sorted lists, merges them into a single sorted list of double the size, and streams the data out a pipe.
+|`pipe_array.hpp`                | Header file containing the definition of an array of pipes.
+|`pipe_array_internal.hpp`       | Helper for pipe_array.hpp.
+|`Produce.hpp`                   | The `Produce` kernel for the merge unit. This kernel reads from input pipes or performs strided reads from device memory and writes the data to an output pipe.
+|`SortingNetworks.hpp`           | Contains the `SortingNetwork` kernel and the `BitonicSortingNetwork` and `MergeSortNetwork` helper functions.
+|`static_math.hpp`               | Miscellaneous static math helper functions.
+|`UnrolledLoop.hpp`              | A templated-based loop unroller that unrolls loops in the compiler front end.
+
+### Merge Sort Details
+This section will describe how the merge sort design is layed out and how it takes advantage of the spatial computing of the FPGA. <br/>
+
+The figure below shows the conceptual view of the merge sort design to the user. The user streams data into a SYCL pipe and, after some delay, the elements are streamed out of the sorter, in sorted order. The number of elements the merge sort design sorts can be a runtime parameter, but it must be a power of 2. However, this restriction can be worked around by padding the input stream with min/max elements, depending on the order of the sort. This technique is demonstrated in this design (see the `fpga_sort` function in *merge_sort.cpp*).
+
+<img src="sort_api.png" alt="sort_api" width="500"/>
+
+The basis of the merge sort design is what we call a *merge unit*, which is shown in the figure below. A single merge unit streams in two sorted lists of size `count` in parallel and merges them into a single sorted list of size `2*count`. The lists are streamed in from device memory (e.g., DDR or HBM) by two `Produce` kernels. The `Consumer` kernel can stream data to either a SYCL pipe or to device memory.
+
+<img src="merge_unit.png" alt="merge_unit" width="600"/>
+
+A single merge unit requires `lg(N)` iterations to sort `N` elements. This requires the host enqueuing `lg(N)` iterations of the merge unit kernels that merge sublists of size {`1`, `2`, `4`, ...} into larger lists of size {`2`, `4`, `8`, ...}, respectively. This results in a timeline that looks like the figure below.
```

Comment:
"This requires the host enqueuing `lg(N)` iterations of the merge unit kernels"
Should it be "merge unit kernel" instead of "kernels" here?

Reply:
Each merge unit has 4 kernels.